tutorials/workshop/Workshop_4_Discriminative_Model_Training.ipynb

```diff
@@ -0,0 +1,286 @@
+   "source": [
+    "## I. Loading Candidates and Gold Labels\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pickle\n",
+    "\n",
+    "with open('new_dev_data.pkl', 'rb') as f:\n",
+    "    dev_data = pickle.load(f)\n",
+    "    dev_labels = pickle.load(f)\n",
+    "    \n",
+    "with open('new_train_data.pkl', 'rb') as f:\n",
+    "    train_data = pickle.load(f)\n",
+    "    train_labels = pickle.load(f)\n",
+    "    \n",
+    "with open('new_test_data.pkl', 'rb') as f:\n",
+    "    test_data = pickle.load(f)\n",
+    "    test_labels = pickle.load(f)\n",
+    "\n",
+    "with open('train_proba.pkl', 'rb') as f:\n",
+    "    train_marginals = pickle.load(f)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## II. Training a _Long Short-term Memory_ (LSTM) Neural Network\n",
+    "\n",
+    "[LSTMs](https://en.wikipedia.org/wiki/Long_short-term_memory) can acheive state-of-the-art performance on many text classification tasks. We'll train a simple LSTM model below. \n",
+    "\n",
+    "In deep learning, hyperparameter tuning is very important and computationally expensive step in training models. For purposes of this tutorial, we have a pre-trained model using the training labels generated from the previous notebook."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Data Processing for LSTM"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First, we prepare the input to our LSTM by adding *markers* to the beginning and end of the person mentions so the LSTM knows which two persons in the sentence we want to learn a relation for. We then featurize the tokens it using a standard vocabulary."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from utils import EmbeddingFeaturizer\n",
+    "from utils import mark_entities\n",
+    "\n",
+    "markers = ['[[BEGIN0]]','[[END0]]','[[BEGIN1]]','[[END1]]']\n",
+    "featurizer = EmbeddingFeaturizer(markers=markers)\n",
+    "\n",
+    "def convert_to_lstm_input(data):\n",
+    "    X = []\n",
+    "    #mark candidates with markers\n",
+    "    for i in range(len(data)):\n",
+    "        cand = data.loc[i]\n",
+    "        marked_tokens = mark_entities(\n",
+    "                    cand.tokens,\n",
+    "                    positions=[cand.person1_word_idx, cand.person2_word_idx],\n",
+    "                    markers=markers)\n",
+    "        X.append(marked_tokens)\n",
+    "        \n",
+    "    #featurize string tokens tokens\n",
+    "    featurize_X = featurizer.fit_transform(X)\n",
+    "    return featurize_X\n",
+    "\n",
+    "train_X_tensor = convert_to_lstm_input(train_data)\n",
+    "dev_X_tensor = convert_to_lstm_input(dev_data)\n",
+    "test_X_tensor = convert_to_lstm_input(test_data)"
```

Comment:
I get a `snorkel.mtl module not found` error, I think we need an `__init__.py` file?

Reply:
Everything needs __init__'s, including the sub-folders in mtl